OvmfPkg: include serial redirection opt by default

macpijan opened the pull request.

macpijan committed:
.github/workflows/build.yml: add

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
OvmfPkg: include serial redirection opt by default

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
OvmfPkg: allow overriding default console redirection state

It is generally useful for testing, to be able to build
QEMU firmware which will have serial enabled by default.

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
.github/scripts/build-qemu.sh: enable serial by default

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
.github/scripts/run-qemu.sh: add

This script can be useful for testing in OSFV.
The serial port can be accessed with:

telnet 127.0.0.1 1234

This way, we can control the edk2 menus in tests.

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan commented:
This MR depends on: https://github.com/Dasharo/edk2/pull/90
I wanted the CI build action to be already included here.

miczyg1 commented on OvmfPkg/OvmfPkgX64.dsc:
```diff
@@ -41,6 +41,8 @@
   DEFINE DISABLE_MTRR_PROGRAMMING = TRUE
   DEFINE IOMMU_ENABLE            = FALSE
   DEFINE SETUP_PASSWORD_ENABLE   = TRUE
+  DEFINE SERIAL_TERMINAL         = TRUE
+  DEFINE SERIAL_REDIRECTION_DEFAULT = FALSE
```
I don't think we need those. We can have it enabled at all times, because if there is a serial port, we always want it enabled I think.

macpijan replied:
@miczyg1 For QEMU and firewall (like Protectli) most likely yes. How about the MSI or laptops? Wouldn't it slow down the boot process and so on?

miczyg1 replied:
@macpijan real hardware does not build OVMF target so not relevant

macpijan replied:
> real hardware does not build OVMF target so not relevant

@miczyg1 

I meant that we might need to expose such option for UefiPayloadPkg.
Removed it for OVMF as reuested. Now the serial redirection is always enabled by default on OFMF, so tests can be executed right after the build. See: https://github.com/Dasharo/edk2/pull/91/commits/55e735decdede473340a025e924d019497b9af15

miczyg1 replied:
SERIAL_TERMINAL is defined in UefiPayloadPkg. Also the default value of the redirection option is handled by PCD. The value of the PCD is passed to the build command by coreboot build system via `--pcd <pcdname>=<value>` so all your concerns are already addressed long time ago.

miczyg1 replied:
https://github.com/Dasharo/coreboot/blob/dasharo/release/payloads/external/edk2/Makefile#L251

macpijan replied:
Thanks @miczyg1 it in fact resolves my concerns. Anything more needed to close this thread?

macpijan committed:
.github/workflows/build.yml: store artifacts

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
OvmfPkg/Library/PlatformBootManagerLib/BdsPlatform.c: print setup keys

Our tests expect these strings to be there, as they are for the
UefiPayloadPkg as well.

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
OvmfPkg: register BootManagerMnuApp

Ported from EmulatorPkg

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan commented:
@miczyg1 Another problem when trying to execute our tests was that the BootManagerMenuApp was not registered (the same Setup Menu was registered on both ESC and F2 keys).

[This commit](https://github.com/Dasharo/edk2/pull/91/commits/115aa9e4aefdde62a99b5cdf59200509acbb8e36) registers the BootManagerMenuApp, so F2 allows to enter it. But it only works when using QEMU in graphical mode. When pressing F2 over serial, it still enters Setup Menu, not Boot Manager App (one time boot).

miczyg1 commented:
> @miczyg1 Another problem when trying to execute our tests was that the BootManagerMenuApp was not registered (the same Setup Menu was registered on both ESC and F2 keys).

F2 was EDK2 default for the setup, so BootManagerMenuApp should rather be ESC.

> 
> [This commit](https://github.com/Dasharo/edk2/pull/91/commits/115aa9e4aefdde62a99b5cdf59200509acbb8e36) registers the BootManagerMenuApp, so F2 allows to enter it. But it only works when using QEMU in graphical mode. When pressing F2 over serial, it still enters Setup Menu, not Boot Manager App (one time boot).

It might be a problem of terminal type and how it parses the escape codes. For UEFIPayload we also had problem with certain Fx keys. The solution was to set the gEfiMdePkgTokenSpaceGuid.PcdDefaultTerminalType to 4.

macpijan committed:
OvmfPkg: always enable serial console redirection

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
OvmfPkg: set terminal type to EFI_TTY_TERM

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan commented:
> F2 was EDK2 default for the setup, so BootManagerMenuApp should rather be ESC.

@miczyg1  Default where? Any reference? Both upstream and mrchromebox just registers Setup Menu on both ESC and F2 AFAIK: https://github.com/MrChromebox/edk2/blob/uefipayload_202309/UefiPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c#L245

We use ESC as setup menu on platforms such as MSI, so it felt natural here as well.

> It might be a problem of terminal type and how it parses the escape codes. For UEFIPayload we also had problem with certain Fx keys. The solution was to set the gEfiMdePkgTokenSpaceGuid.PcdDefaultTerminalType to 4.

Changed in: https://github.com/Dasharo/edk2/pull/91/commits/3fd08d41cd48d8273752ace0caab6ba91c9996e9

Now F2 enters one time boot menu properly over serial (telnet).

macpijan committed:
OvmfPkg: sync the UEFI Shel entry name with UefiPayload

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

miczyg1 commented:
> > F2 was EDK2 default for the setup, so BootManagerMenuApp should rather be ESC.
> 
> @miczyg1 Default where? Any reference? Both upstream and mrchromebox just registers Setup Menu on both ESC and F2 AFAIK: https://github.com/MrChromebox/edk2/blob/uefipayload_202309/UefiPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c#L245

Ahh that was very long time ago, like in the beginning of usage of EDK2 by us. It was F2 and arrow down IIRC so even more different from what is now: https://github.com/Dasharo/edk2/commit/9dd88d88fc688b7a9a320740534d04c6da335fae

 MrChromebox has it modified as well.

> 
> We use ESC as setup menu on platforms such as MSI, so it felt natural here as well.

All platforms that are based on AMI (Protectli, MSI)  use DEL for setup and F11 for boot manager menu.

macpijan commented:
@miczyg1 So what bindings shall be used here so we can proceed with this MR?

Esc- one time boot
F2 - setup

?

miczyg1 commented:
> @miczyg1 So what bindings shall be used here so we can proceed with this MR?
> 
> Esc- one time boot F2 - setup
> 
> ?

@macpijan exactly as above would be great. it will also somewhat match the Insyde/NovaCustom setup hotkey.

macpijan committed:
OvmfPkg: swap Esc and F2 mapping

We want the following:
- ESC - one time boot menu
- F2 - setup menu

See the discussion in Github:
https://github.com/Dasharo/edk2/pull/91#issuecomment-1757624392

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
.github/scripts/run-qemu.sh: run with graphic ON by default

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan commented:
@miczyg1 Changed keys mapping in: https://github.com/Dasharo/edk2/pull/91/commits/5316967ff561a6743d44eeab435ff4d6a55f322a

And rebased on top of the recent github-actions branch, which should be merged first.

miczyg1 commented on OvmfPkg/Library/PlatformBootManagerLib/BdsPlatform.c:
```diff
@@ -1475,6 +1628,12 @@ PlatformBootManagerAfterConsole (
+  // This is probably the earliest we can print this, as before the console is
+  // not ready yet.
+  Print(L"F2 to enter Setup\n");
+  Print(L"Esc to enter Boot Manager Menu\n");
```
```suggestion
  Print(L"ESC to enter Boot Manager Menu\n");
```

miczyg1 committed:
Update OvmfPkg/Library/PlatformBootManagerLib/BdsPlatform.c

miczyg1 commented on OvmfPkg/OvmfPkgX64.fdf:
```diff
@@ -257,7 +257,7 @@ INF  MdeModulePkg/Universal/CapsuleRuntimeDxe/CapsuleRuntimeDxe.inf
 INF  MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitterDxe.inf
 INF  MdeModulePkg/Universal/Console/GraphicsConsoleDxe/GraphicsConsoleDxe.inf
-INF  MdeModulePkg/Universal/Console/TerminalDxe/TerminalDxe.inf
+INF MdeModulePkg/Universal/Console/TerminalDxe/TerminalDxe.inf
```
```suggestion
INF  MdeModulePkg/Universal/Console/TerminalDxe/TerminalDxe.inf
```

miczyg1 committed:
Update OvmfPkg/OvmfPkgX64.fdf
